download font from url on build

landonr opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3261

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
external_components:
  - source:
      type: local
      path: ../compile-online-image/esphome/components
    components: [
      font,
      external_files
    ]

font:
  - file: 
      url: "https://github.com/IdreesInc/Monocraft/releases/download/v3.0/Monocraft.ttf"
      # refresh: 10s
      type: web
    id: web_font
    size: 24
    glyphs: '^…*/|\$#<>!?"%()[]+=,-_.:°0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz&êÊîÎâÂôÔéÉíÍáÁóÓčëËüÜïÏöÖøØṣṢñãõÑÃÕ''’'

  - file: 
      url: "https://github.com/IdreesInc/Monocraft/releases/download/v3.0/Monocraft.ttf"
      # refresh: 10s
      type: web
    id: web_font_medium
    size: 20
    glyphs: '^…*/|\$#<>!?"%()[]+=,-_.:°0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz&êÊîÎâÂôÔéÉíÍáÁóÓčëËüÜïÏöÖøØṣṢñãõÑÃÕ''’'

  - file: "fonts/iosevka.ttf"
    id: local_font
    size: 16
    glyphs: '^…*/|\$#<>!?"%()[]+=,-_.:°0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz&êÊîÎâÂôÔéÉíÍáÁóÓčëËüÜïÏöÖøØṣṢñãõÑÃÕ''’'
  - file: "https://github.com/landonr/homeThing/raw/main/fonts/materialdesignicons-webfont.ttf" 
    id: material_font_large
    size: 20
    glyphs: [
      '󰐊', # mdi-play
    ]
  - file: "gfonts://Roboto"
    id: roboto
    size: 20


spi:
  clk_pin: GPIO18
  mosi_pin: GPIO19

display:
  - platform: st7789v
    model: TTGO_TDisplay_135x240
    id: my_display
    cs_pin: GPIO5
    dc_pin: GPIO16
    reset_pin: GPIO23
    rotation: 90
    update_interval: 10s
    lambda: |-
      it.print(0, 0, id(web_font), "web font");
      it.print(0, 25, id(web_font_medium), "web font medium");
      it.print(0, 50, id(local_font), "local font");
      it.print(0, 75, id(material_font_large), "󰐊");
      it.print(0, 100, id(roboto), "gfont");
      return;
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

![image](https://github.com/esphome/esphome/assets/2607659/5ae40ccc-6065-4582-9888-9fef3190988e)

landonr committed:
load font from url

landonr committed:
black format

landonr committed:
pylint format

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -109,9 +110,42 @@ def _compute_gfonts_local_path(value) -> Path:
+    return None
+
+
+def get_file_type(file_name):
+    # Regular expression pattern to match the file extension
+    pattern = r"\.([^.]+)$"
+    match = re.search(pattern, file_name)
+    if match:
+        file_type = match.group(1)
+        return file_type
+    return None
```
As in #5214, this could be combined into the get_filename_from_url method

landonr replied:
i don't really understand this comment, can you suggest the change? How would the method return 2 different options

guillempages replied:
You can return a tuple:
```
return (file_name, file_type)
```

and when calling the method you can do
```
  filename, extension = get_file_from_url(url)
```
And you would fill both variables with a single call. Python magic ;-)

landonr replied:
got it working, thanks for the suggestion!

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -109,9 +110,42 @@ def _compute_gfonts_local_path(value) -> Path:
+    pattern = r"/([^/]+)$"
+    match = re.search(pattern, url)
+    if match:
+        file_name = match.group(1)
```
You might want to rename to filename for consistency with the method name
```suggestion
        filename = match.group(1)
```

landonr replied:
i renamed the method instead because file_name was used elsewhere already

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -109,9 +110,42 @@ def _compute_gfonts_local_path(value) -> Path:
+    match = re.search(pattern, url)
+    if match:
+        file_name = match.group(1)
+        file_name = re.sub(r"%20", "", file_name)  # Replace %20 with empty string
```
"%20" is normally a space; maybe replace by "_" for better readability?

landonr replied:
done!

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -191,6 +225,35 @@ def download_gfonts(value):
 )
 
 
+def download_font(value):
```
Maybe combine with `download_gfont`? (i.e. write the common functionality only once and keep the methods only for parsing the URLs/names)

landonr replied:
good idea, I had to call it from the higher level config, but it looks a lot cleaner

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -207,6 +270,14 @@ def validate_file_shorthand(value):
             data[CONF_WEIGHT] = weight[1:]
         return FILE_SCHEMA(data)
 
+    if value.startswith("https"):
```
unencrypted http is not fully dead yet ;-)
```suggestion
    if value.startswith("http://") or value.startswith("https://"):
```

landonr replied:
long live http. what do you suggest? I can look for a regex to match URLs if you think that works?

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -109,9 +110,42 @@ def _compute_gfonts_local_path(value) -> Path:
+    return None
+
+
+def _compute_local_image_path(value) -> Path:
```
Could this be combined with the gfont one?
Also, the hash calculation used for gfonts might also be useful here, to allow downloading diffenent fonts with the same name (but different URLs)

landonr replied:
I'll check back on this. I combined the methods but I might be able to fix the hashing to support different fonts with the same name

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -364,6 +436,9 @@ async def to_code(config):
         path = _compute_gfonts_local_path(conf)
         font = load_ttf_font(path, config[CONF_SIZE])
+    elif conf[CONF_TYPE] == TYPE_WEB:
+        path = _compute_local_image_path(conf).as_posix()
```
Could the `as_posix()` call be done within the `compute_path` method? (this comment might become irrelevant if you do combine it with the `_compute_gfonts_local_path` method).

landonr replied:
I removed this, but I haven't tested it yet

landonr commented:
thanks for the review @guillempages i'll do the changes this week. It might make sense to move out the `compute_local_path` stuff to a new file because it can be used between fonts and images.

I've also been meaning to add an option for `refresh: 1day` so images can be updated

landonr committed:
combine download_font and download_gfont

landonr committed:
comboine compute_local_path stuff into get_font_path

landonr committed:
move download web font to a less generic place

landonr committed:
pylint and format

landonr committed:
support http fonts

landonr committed:
move external file code into it's own component

landonr committed:
format

landonr committed:
fix gfont

landonr commented:
> thanks for the review @guillempages i'll do the changes this week. It might make sense to move out the `compute_local_path` stuff to a new file because it can be used between fonts and images.
> 
> I've also been meaning to add an option for `refresh: 1day` so images can be updated

@guillempages i updated the external_files stuff to a new component called `external_files`. I also added the `source_refresh` key, but made it default to `never`

landonr committed:
remove unused import

landonr committed:
Merge branch 'dev' into online_font

landonr committed:
add codeowners

landonr committed:
updated codeowners file

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        return recent_directory
+
+    current_time = int(datetime.now().timestamp())
+    sub_directory = Path(os.path.join(base_directory, str(current_time)))
```
I don't think it's a good idea to hard-code the timestamp into the directory name. Both Linux and Windows filesystems do store metadata related to the file like when was it created; no need to make this explicit in the directory name.

landonr replied:
i thought about that but I wasn't sure if that was a solid solution. It's an easy switch if you think that's better!

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+    return None
+
+
+def compute_local_file_dir(name: str, domain: str, refresh: TimePeriodSeconds) -> Path:
```
IMHO "refresh" would only be relevant when deciding whether to re-download or not, but the whole directory creation methods should be agnostic to the refresh time.

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+def is_directory_recent(directory_path: str, refresh: TimePeriodSeconds) -> bool:
+    if os.path.exists(directory_path):
+        current_time = datetime.now().timestamp()
+        dir_timestamp = int(os.path.basename(directory_path))
```
You should read the timestamp from the file stats; not from the directory name.

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+    return False
+
+
+def find_recent_directory(base_directory: str, refresh: TimePeriodSeconds) -> Path:
```
I'm not sure this method is needed at all (I know you are calling it; I mean I don't think this should be called ;-) )

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+    if not os.path.exists(base_directory):
+        os.makedirs(base_directory)
+
+    recent_directory = find_recent_directory(base_directory, refresh)
+    if recent_directory:
+        return recent_directory
```
The directory should be unrelated to the timestamp.

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+
+def get_file_name_from_url(url):
+    # Regular expression pattern to match the file name at the end of the URL
+    pattern = r"/([^/]+)$"
```
As mentioned before, you could combine "filename" and "extension" into a single regex, to optimize the code a bit. If you then `return (filename, extension)`, callers can get the split for free...

guillempages commented on esphome/components/external_files/__init__.py:
```diff
@@ -0,0 +1,72 @@
+    return None
+
+
+def get_file_type(file_name):
```
...And then you wouldn't need this method.

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -8,6 +8,7 @@
 import requests
 
 from esphome import core
+from esphome.components import external_files
```
I'm not sure if this is actually a component, or would be a utility; but that's not for me to decide; that should be decided by the project owners (Jesse & Co)

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -142,25 +134,48 @@ def validate_weight_name(value):
+        )
+    if value[CONF_TYPE] == TYPE_WEB:
+        return value[CONF_URL]
+    return ""
```
Maybe better return None, instead of an empty string?
```suggestion
    return None
```

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -142,25 +134,48 @@ def validate_weight_name(value):
+        return f"{value[CONF_FAMILY]}@{value[CONF_WEIGHT]}"
+    if value[CONF_TYPE] == TYPE_WEB:
+        return external_files.get_file_name_from_url(value[CONF_URL])
+    return ""
```
```suggestion
    return None
```

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -142,25 +134,48 @@ def validate_weight_name(value):
+        name = f"{value[CONF_FAMILY]}@{value[CONF_WEIGHT]}@{value[CONF_ITALIC]}@v1"
+        return (
+            Path(
+                external_files.compute_local_file_dir(name, DOMAIN, value[CONF_REFRESH])
```
Previously, the directory contained a hashed version of the font name; now it contains the name in plaintext. I don't have a strong opinion either way, but I assume the previous implementation was done this way for a reason; maybe Jesse does have some input on that.

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -142,25 +134,48 @@ def validate_weight_name(value):
+            external_files.compute_local_file_dir(name, DOMAIN, value[CONF_REFRESH])
+        )
+        return file_path / file_name
+    return ""
```
```suggestion
    return None
```

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -142,25 +134,48 @@ def validate_weight_name(value):
+            / "font.ttf"
         )
+    if value[CONF_TYPE] == TYPE_WEB:
+        font_id = external_files.get_file_name_from_url(value[CONF_URL])
```
You created the method "get_font_name", but are not using that?

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -173,23 +188,50 @@ def download_gfonts(value):
+    if path.is_file():
+        return value
+    try:
+        req = requests.get(url, timeout=30)
```
Maybe create a constant for the timeout?

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -173,23 +188,50 @@ def download_gfonts(value):
         cv.Optional(CONF_ITALIC, default=False): cv.boolean,
-    },
-    download_gfonts,
+        cv.Optional(CONF_REFRESH, default="never"): cv.All(
```
Not sure about the "refresh" value. Maybe a more automatic system, using the ETAG header from the HTTP request would be better? If the server properly implements that, it would return an id on the etag header, that would indicate whether the file changed on the server or not. (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag)

landonr replied:
interesting, i'll check that out! I made it default to never because it could return a new directory if the build took over a second. This should solve that

landonr replied:
how do you think we should keep track of the etag? In the file name or directory?

guillempages commented on esphome/components/font/__init__.py:
```diff
@@ -254,10 +308,13 @@ def _file_schema(value):
 )
 
-CONFIG_SCHEMA = cv.All(validate_pillow_installed, FONT_SCHEMA)
+CONFIG_SCHEMA = cv.All(
```
Was this change necessary because of the CI formatting checks (black, pyflake,...)?

landonr replied:
I forgot to remove this, thanks!

landonr committed:
Update esphome/components/font/__init__.py

Co-authored-by: guillempages <guillempages@users.noreply.github.com>

landonr commented:
thanks @guillempages i'll get to work on this later today or tomorrow

landonr committed:
Merge branch 'dev' into online_font

landonr committed:
use file creation metadata instead of including it in the file names

landonr committed:
undo extra comma

landonr committed:
Merge branch 'dev' of github.com:esphome/esphome into online_font

# Conflicts:
#	esphome/components/font/__init__.py

landonr committed:
fix downloading gfont

landonr committed:
Merge branch 'dev' into online_font

landonr commented on esphome/components/external_files/__init__.py:
I moved this to a new file so we can reuse it for images

jesserockz replied:
This is probably better off in a helper/util file, not a component file

landonr replied:
I'm going to reuse it for external images, i'll move it out to a helper file

landonr committed:
add web font to test

landonr committed:
fix whitespace

landonr committed:
Merge branch 'dev' into online_font

landonr committed:
Merge branch 'dev' into online_font

landonr committed:
move external files to a helper file

landonr committed:
fix import

landonr committed:
codeowners

landonr committed:
unquote file names

landonr committed:
Merge branch 'devd' into online_font

# Conflicts:
#	esphome/components/font/__init__.py

landonr committed:
Update esphome/components/font/__init__.py

Co-authored-by: guillempages <guillempages@users.noreply.github.com>

landonr committed:
Update esphome/components/font/__init__.py

Co-authored-by: guillempages <guillempages@users.noreply.github.com>

landonr committed:
use etag for refresh

landonr commented on esphome/components/font/__init__.py:
```diff
@@ -162,27 +179,58 @@ def download_gfonts(value):
+    if external_files.is_file_recent(
+        path, value[CONF_REFRESH]
+    ) or external_files.check_etag_equality(url, path):
+        return value
```
return the font if the file is recent, or the local etag matches the remote

landonr committed:
use last modified date instead of etag

landonr committed:
switch if statement to not

landonr committed:
Merge branch 'devd' into online_font

landonr committed:
Merge branch 'devd' into online_font

landonr committed:
Merge branch 'devd' into online_font

# Conflicts:
#	esphome/external_files.py

landonr committed:
add logging

landonr committed:
use updated external_files

landonr committed:
Merge branch 'devd' into online_font

landonr committed:
fix gfonts not working

landonr commented:
@jesserockz updated with changes from #5214

landonr committed:
Merge branch 'dev' into online_font

jesserockz commented on esphome/components/font/__init__.py:
```diff
@@ -221,17 +277,23 @@ def validate_file_shorthand(value):
         return validate_file_shorthand(value)
-    return TYPED_FILE_SCHEMA(value)
+    typed_schema = TYPED_FILE_SCHEMA(value)
+    if typed_schema[CONF_TYPE] == TYPE_WEB:
+        download_web_font(typed_schema)
+    elif typed_schema[CONF_TYPE] == TYPE_GFONTS:
+        download_gfont(typed_schema)
```
I think these should still be on the relevant SCHEMAs as extras...

landonr replied:
Thanks! Updated, let me know what you think

landonr committed:
Merge branch 'devd' into online_font

landonr committed:
move download calls to schema

jesserockz commented on esphome/components/font/__init__.py:
```diff
@@ -161,26 +200,45 @@ def download_gfonts(value):
+    }
+)
+
+CONF_DOWNLOAD = "download"
+
+GFONTS_SCHEMA = EXTERNAL_FONT_SCHEMA.extend(
+    {
+        cv.Required(CONF_FAMILY): cv.string_strict,
+        cv.Optional(CONF_DOWNLOAD): download_gfont,
+    }
+).extend(EXTERNAL_FONT_SCHEMA)
+
+WEB_FONT_SCHEMA = EXTERNAL_FONT_SCHEMA.extend(
+    {
+        cv.Required(CONF_URL): cv.string_strict,
+        cv.Optional(CONF_DOWNLOAD): download_web_font,
+    }
 )
```
No, it should be like this I think.
```suggestion
GFONTS_SCHEMA = cv.All(
    EXTERNAL_FONT_SCHEMA.extend(
        {
            cv.Required(CONF_FAMILY): cv.string_strict,
        }
    ),
    download_gfont,
)

WEB_FONT_SCHEMA = cv.All(
    EXTERNAL_FONT_SCHEMA.extend(
        {
            cv.Required(CONF_URL): cv.string_strict,
        }
    ),
    download_web_font,
)
```

landonr replied:
nice I couldn't figure out how to get the cv.All to have a schema. I had to pass in the type because it wasn't available in the conf
Updated, hopefully it looks good!

landonr committed:
update schemas using suggestions

landonr committed:
Merge branch 'devd' into online_font

landonr committed:
Merge branch 'devd' into online_font

jesserockz committed:
Merge branch 'dev' into pr/landonr/5254

jesserockz committed:
Remove unused const

jesserockz committed:
No need to save as local variable

clydebarrow committed:
Move test to tests/components/font/

clydebarrow committed:
Blank line removed

clydebarrow reviewed (approved):
Tested the earlier font changes with this, all seems well.

jesserockz commented:
Thanks @clydebarrow. Im just fixing up the docs PR now
